File: examples/replicate-fig2davide.cc
/*
 * This script attempts to replicate Fig 2 of Davide's A Thorough Study of LoRaWAN Performance 
 *  Under Different Parameter Settings paper.
 * 
 * Based off complete-network-example.cc
 * Sim setup given below
 * No shadowing
 * Trying 100% confirmed, 100% unconfirmed and 50/50 traffic
 * num of trans = 8
 * 1200 devices (from Martina's email)
 * data rate sweeped from 10^-2 to 2 pkt/s by sem script
 * radius 6300m (from email with Martina)
 * 1 gateway
 * simulation time = 3*appPeriod (from Martina's email)
 * Stat collection is only done on the second period's time interval (from Martina's email)
 * no ADR, using set SF up
 * packet size a uniform random value between 0 and 40 (from Martina's email)
 * GW DC enabled (this is so by default in gateway-lorawan-mac.cc)
 * TX is priorities (this is so by default m_txPriority in lora-phy-helper.h)
 * Sub-band prioritization is Off (by default, sim needs additional code to enable this)
 * RX2 data rate is SF12 (LorawanMacHelper setup for EU function sets this)
 * Number of TX attempts, m = 8 (set in this script  edLorawanMac->SetMaxNumberOfTransmissions (numberOfTransmissions);)
 * Full duplex = No (by default, sim needs additional code for this)
 * Reception paths = 8 (LorawanMacHelper setup for EU function sets this)
 * 
 * 
 *  *  How to run and save output in a file
 * ./waf configure --enable-tests --enable-examples
 * ./waf build
 * ./waf --run "replicate-fig2davide --confirmed=True --appPeriod=120000" > output.c 2>&1
 */




#include "ns3/end-device-lora-phy.h"
#include "ns3/gateway-lora-phy.h"
#include "ns3/class-a-end-device-lorawan-mac.h"
#include "ns3/gateway-lorawan-mac.h"
#include "ns3/simulator.h"
#include "ns3/log.h"
#include "ns3/pointer.h"
#include "ns3/constant-position-mobility-model.h"
#include "ns3/lora-helper.h"
#include "ns3/node-container.h"
#include "ns3/mobility-helper.h"
#include "ns3/position-allocator.h"
#include "ns3/double.h"
#include "ns3/random-variable-stream.h"
#include "ns3/periodic-sender-helper.h"
#include "ns3/command-line.h"
#include "ns3/network-server-helper.h"
#include "ns3/correlated-shadowing-propagation-loss-model.h"
#include "ns3/building-penetration-loss.h"
#include "ns3/building-allocator.h"
#include "ns3/buildings-helper.h"
#include "ns3/forwarder-helper.h"
#include <algorithm>
#include <ctime>

using namespace ns3;
using namespace lorawan;

NS_LOG_COMPONENT_DEFINE ("ReplicateDavideFig2");

// Network settings
int nDevices = 1200;
int nGateways = 1;
double radius = 6300;
int appPeriodSeconds = 44; //Will be changed by sem script
int confirmedPercentage = 100;
double simulationTime = 3*appPeriodSeconds; //Will be overwritten below (line 96)
bool confirmedTraffic = true;

// Channel model
//Paper only used the log-distance parts (no shadowing)
bool realisticChannelModel = false;


uint8_t numberOfTransmissions = 8; // The maximum number of transmissions allowed

// Output control
bool print = true;

int
main (int argc, char *argv[])
{

  CommandLine cmd;
  cmd.AddValue ("appPeriod",
                "The period in seconds to be used by periodically transmitting applications",
                appPeriodSeconds);
  cmd.AddValue("confirmedPercentage", "Which percentage of devices should employ confirmed packets", confirmedPercentage);
  //cmd.AddValue ("confirmed", "Whether or not end devices require an ACK", confirmedTraffic);
  cmd.Parse (argc, argv);

  simulationTime = 3*appPeriodSeconds; //Update sim time with new value

  // Set up logging
  LogComponentEnable ("ReplicateDavideFig2", LOG_LEVEL_ALL);
  // LogComponentEnable("LoraChannel", LOG_LEVEL_INFO);
  // LogComponentEnable("LoraPhy", LOG_LEVEL_ALL);
  // LogComponentEnable("EndDeviceLoraPhy", LOG_LEVEL_ALL);
  // LogComponentEnable("GatewayLoraPhy", LOG_LEVEL_ALL);
  // LogComponentEnable("LoraInterferenceHelper", LOG_LEVEL_ALL);
  // LogComponentEnable("LorawanMac", LOG_LEVEL_ALL);
  // LogComponentEnable("EndDeviceLorawanMac", LOG_LEVEL_ALL);
  // LogComponentEnable("ClassAEndDeviceLorawanMac", LOG_LEVEL_ALL);
  // LogComponentEnable("GatewayLorawanMac", LOG_LEVEL_ALL);
  // LogComponentEnable("LogicalLoraChannelHelper", LOG_LEVEL_ALL);
  // LogComponentEnable("LogicalLoraChannel", LOG_LEVEL_ALL);
  // LogComponentEnable("LoraHelper", LOG_LEVEL_ALL);
  // LogComponentEnable("LoraPhyHelper", LOG_LEVEL_ALL);
  // LogComponentEnable("LorawanMacHelper", LOG_LEVEL_ALL);
  // LogComponentEnable("PeriodicSenderHelper", LOG_LEVEL_ALL);
  // LogComponentEnable("PeriodicSender", LOG_LEVEL_ALL);
  // LogComponentEnable("LorawanMacHeader", LOG_LEVEL_ALL);
  // LogComponentEnable("LoraFrameHeader", LOG_LEVEL_ALL);
  // LogComponentEnable("NetworkScheduler", LOG_LEVEL_ALL);
  // LogComponentEnable("NetworkServer", LOG_LEVEL_ALL);
  // LogComponentEnable("NetworkStatus", LOG_LEVEL_ALL);
  // LogComponentEnable("NetworkController", LOG_LEVEL_ALL);
  //LogComponentEnable("LoraPacketTracker", LOG_LEVEL_ALL);
  
  LogComponentEnableAll (LOG_PREFIX_FUNC);
  LogComponentEnableAll (LOG_PREFIX_NODE);
  LogComponentEnableAll (LOG_PREFIX_TIME);
  /***********
   *  Setup  *
   ***********/

  // Create the time value from the period
  Time appPeriod = Seconds (appPeriodSeconds);

  // Mobility
  MobilityHelper mobility;
  mobility.SetPositionAllocator ("ns3::UniformDiscPositionAllocator", "rho", DoubleValue (radius),
                                 "X", DoubleValue (0.0), "Y", DoubleValue (0.0));
  mobility.SetMobilityModel ("ns3::ConstantPositionMobilityModel");

  /************************
   *  Create the channel  *
   ************************/

  // Create the lora channel object
  Ptr<LogDistancePropagationLossModel> loss = CreateObject<LogDistancePropagationLossModel> ();
  loss->SetPathLossExponent (3.76);
  loss->SetReference (1, 8.1); //used to be 7.7. 8.1 comes from private lorawan repo's rawcompletenetwork example

  if (realisticChannelModel)
    {
      // Create the correlated shadowing component
      Ptr<CorrelatedShadowingPropagationLossModel> shadowing =
          CreateObject<CorrelatedShadowingPropagationLossModel> ();

      // Aggregate shadowing to the logdistance loss
      loss->SetNext (shadowing);

      // Add the effect to the channel propagation loss
      Ptr<BuildingPenetrationLoss> buildingLoss = CreateObject<BuildingPenetrationLoss> ();

      shadowing->SetNext (buildingLoss);
    }

  Ptr<PropagationDelayModel> delay = CreateObject<ConstantSpeedPropagationDelayModel> ();

  Ptr<LoraChannel> channel = CreateObject<LoraChannel> (loss, delay);

  /************************
   *  Create the helpers  *
   ************************/

  // Create the LoraPhyHelper
  LoraPhyHelper phyHelper = LoraPhyHelper ();
  phyHelper.SetChannel (channel);

  // Create the LorawanMacHelper
  LorawanMacHelper macHelper = LorawanMacHelper ();

  // Create the LoraHelper
  LoraHelper helper = LoraHelper ();
  helper.EnablePacketTracking (); // Output filename
  // helper.EnableSimulationTimePrinting ();

  //Create the NetworkServerHelper
  NetworkServerHelper nsHelper = NetworkServerHelper ();

  //Create the ForwarderHelper
  ForwarderHelper forHelper = ForwarderHelper ();

  /************************
   *  Create End Devices  *
   ************************/

  // Create a set of nodes
  NodeContainer endDevices;
  endDevices.Create (nDevices);

  // Assign a mobility model to each node
  mobility.Install (endDevices);

  // Make it so that nodes are at a certain height > 0
  for (NodeContainer::Iterator j = endDevices.Begin (); j != endDevices.End (); ++j)
    {
      Ptr<MobilityModel> mobility = (*j)->GetObject<MobilityModel> ();
      Vector position = mobility->GetPosition ();
      position.z = 1.2;
      mobility->SetPosition (position);
    }

  // Create the LoraNetDevices of the end devices
  uint8_t nwkId = 54;
  uint32_t nwkAddr = 1864;
  Ptr<LoraDeviceAddressGenerator> addrGen =
      CreateObject<LoraDeviceAddressGenerator> (nwkId, nwkAddr);

  // Create the LoraNetDevices of the end devices
  macHelper.SetAddressGenerator (addrGen);
  macHelper.SetRegion (LorawanMacHelper::EU);
  phyHelper.SetDeviceType (LoraPhyHelper::ED);
  macHelper.SetDeviceType (LorawanMacHelper::ED_A);
  helper.Install (phyHelper, macHelper, endDevices);

  // Now end devices are connected to the channel

  // Make traffic confirmed if needed

  // Figure out how many devices should employ confirmed traffic
  int confirmedNumber = confirmedPercentage * endDevices.GetN () / 100;
  int i = 0;

  for (NodeContainer::Iterator j = endDevices.Begin (); j != endDevices.End (); ++j)
    {
      Ptr<Node> node = *j;

      // Set message type (default is unconfirmed)
      Ptr<LorawanMac> edMac =node->GetDevice (0)->GetObject<LoraNetDevice> ()->GetMac ();
      Ptr<ClassAEndDeviceLorawanMac> edLorawanMac = edMac->GetObject<ClassAEndDeviceLorawanMac> ();

      // Set message type, otherwise the NS does not send ACKs
      if (i < confirmedNumber)
      {
       edLorawanMac->SetMType (LorawanMacHeader::CONFIRMED_DATA_UP);
      }

      edLorawanMac->SetMaxNumberOfTransmissions (numberOfTransmissions);

    }
  


  /*********************
   *  Create Gateways  *
   *********************/

  // Create the gateway nodes (allocate them uniformely on the disc)
  NodeContainer gateways;
  gateways.Create (nGateways);

  Ptr<ListPositionAllocator> allocator = CreateObject<ListPositionAllocator> ();
  // Make it so that nodes are at a certain height > 0
  allocator->Add (Vector (0.0, 0.0, 15.0));
  mobility.SetPositionAllocator (allocator);
  mobility.Install (gateways);

  // Create a netdevice for each gateway
  phyHelper.SetDeviceType (LoraPhyHelper::GW);
  macHelper.SetDeviceType (LorawanMacHelper::GW);
  helper.Install (phyHelper, macHelper, gateways);

  /**********************
   *  Handle buildings  *
   **********************/

  double xLength = 130;
  double deltaX = 32;
  double yLength = 64;
  double deltaY = 17;
  int gridWidth = 2 * radius / (xLength + deltaX);
  int gridHeight = 2 * radius / (yLength + deltaY);
  if (realisticChannelModel == false)
    {
      gridWidth = 0;
      gridHeight = 0;
    }
  Ptr<GridBuildingAllocator> gridBuildingAllocator;
  gridBuildingAllocator = CreateObject<GridBuildingAllocator> ();
  gridBuildingAllocator->SetAttribute ("GridWidth", UintegerValue (gridWidth));
  gridBuildingAllocator->SetAttribute ("LengthX", DoubleValue (xLength));
  gridBuildingAllocator->SetAttribute ("LengthY", DoubleValue (yLength));
  gridBuildingAllocator->SetAttribute ("DeltaX", DoubleValue (deltaX));
  gridBuildingAllocator->SetAttribute ("DeltaY", DoubleValue (deltaY));
  gridBuildingAllocator->SetAttribute ("Height", DoubleValue (6));
  gridBuildingAllocator->SetBuildingAttribute ("NRoomsX", UintegerValue (2));
  gridBuildingAllocator->SetBuildingAttribute ("NRoomsY", UintegerValue (4));
  gridBuildingAllocator->SetBuildingAttribute ("NFloors", UintegerValue (2));
  gridBuildingAllocator->SetAttribute (
      "MinX", DoubleValue (-gridWidth * (xLength + deltaX) / 2 + deltaX / 2));
  gridBuildingAllocator->SetAttribute (
      "MinY", DoubleValue (-gridHeight * (yLength + deltaY) / 2 + deltaY / 2));
  BuildingContainer bContainer = gridBuildingAllocator->Create (gridWidth * gridHeight);

  BuildingsHelper::Install (endDevices);
  BuildingsHelper::Install (gateways);

  // Print the buildings
  if (print)
    {
      std::ofstream myfile;
      myfile.open ("buildings.txt");
      std::vector<Ptr<Building>>::const_iterator it;
      int j = 1;
      for (it = bContainer.Begin (); it != bContainer.End (); ++it, ++j)
        {  
          Box boundaries = (*it)->GetBoundaries ();
          myfile << "set object " << j << " rect from " << boundaries.xMin << "," << boundaries.yMin
                 << " to " << boundaries.xMax << "," << boundaries.yMax << std::endl;
        }
      myfile.close ();
    }

  /**********************************************
   *  Set up the end device's spreading factor  *
   **********************************************/

  macHelper.SetSpreadingFactorsUp (endDevices, gateways, channel);

  NS_LOG_DEBUG ("Completed configuration");

  /*********************************************
   *  Install applications on the end devices  *
   *********************************************/

  Time appStopTime = Seconds (simulationTime);
  PeriodicSenderHelper appHelper = PeriodicSenderHelper ();
  appHelper.SetPeriod (Seconds (appPeriodSeconds));


  appHelper.SetPacketSize(0);
  Ptr<RandomVariableStream> rv = CreateObjectWithAttributes<UniformRandomVariable> (
      "Min", DoubleValue (0), "Max", DoubleValue (40)); //0 and 40 included

  appHelper.SetPacketSizeRandomVariable(rv); //As indicated by email from Martina

  ApplicationContainer appContainer = appHelper.Install (endDevices);

  appContainer.Start (Seconds (0));
  appContainer.Stop (appStopTime);

  /**************************
   *  Create Network Server  *
   ***************************/

  // Create the NS node
  NodeContainer networkServer;
  networkServer.Create (1);

  // Create a NS for the network
  nsHelper.SetEndDevices (endDevices);
  nsHelper.SetGateways (gateways);
  nsHelper.Install (networkServer);

  //Create a forwarder for each gateway
  forHelper.Install (gateways);

  ////////////////
  // Simulation //
  ////////////////

  Simulator::Stop (appStopTime + Hours (1));

  NS_LOG_INFO ("Running simulation...");
  Simulator::Run ();

  Simulator::Destroy ();

  ///////////////////////////
  // Print results to file //
  ///////////////////////////
  NS_LOG_INFO ("Computing performance metrics...");
  NS_LOG_INFO ("Computing over the period "<< Seconds(appPeriodSeconds)<< " to "<< Seconds(2*appPeriodSeconds));
  //Want to compute using the middle time section of each simulation.
  //Each sim ran for 3*appPeriodSeconds, so computation must use the middle interval
  //PrintPhyPacketsPerGw can take a start and stop time
  LoraPacketTracker &tracker = helper.GetPacketTracker ();

 
      std::cout << tracker.CountMacPacketsGlobally (Seconds (appPeriodSeconds), Seconds(2*appPeriodSeconds)) << std::endl;
      std::cout << tracker.PrintPhyPacketsPerGw (Seconds (appPeriodSeconds), Seconds(2*appPeriodSeconds), nDevices) << std::endl;
    
  return 0;
}

File: examples/replicate_fig2davide.py
"""
/*
 * This script attempts to replicate Fig 2 of Davide's A Thorough Study of LoRaWAN Performance 
 *  Under Different Parameter Settings paper.
 * 
 * Sim setup
 * No shadowing
 * Trying 100% confirmed and  100% unconfirmed traffic
 * num of trans = 8 (confirmed traffic)
 * 1200 devices (from Martina's email)
 * data rate sweeped from 10^-2 to 2 pkt/s by script
 * radius 6300m (from email with Martina)
 * 1 gateway
 * simulation time = 3*appPeriod (from Martina's email)
 * Stat collection is only done on the second period's time interval (from Martina's email)
 * no ADR, using set SF up
 * packet size a uniform random value between 0 and 40 (from Martina's email)
 * GW DC enabled (this is so by default in gateway-lorawan-mac.cc)
 * TX is priorities (this is so by default m_txPriority in lora-phy-helper.h)
 * Sub-band prioritization is Off (by default, sim needs additional code to enable this)
 * RX2 data rate is SF12 (LorawanMacHelper setup for EU function sets this)
 * Number of TX attempts, m = 8 (set in this script  edLorawanMac->SetMaxNumberOfTransmissions (numberOfTransmissions);)
 * Full duplex = No (by default, sim needs additional code for this)
 * Reception paths = 8 (LorawanMacHelper setup for EU function sets this)
 */
"""

import sem
import numpy as np
import matplotlib.pyplot as plt


# Create our SEM campaign
script = 'replicate-fig2davide'
results_dir = 'newreference-results'

#CountMacPacketsGloballyCpsr is saved under cspr
#CountMacPacketsGlobally in saved under glob

ns_3_dir = '../../../'

campaign = sem.CampaignManager.new(ns_3_dir, script, results_dir,
                                   check_repo=False, overwrite=False)

# Parameter space
#################


#The period in seconds to be used by periodically transmitting applications
appPeriod_values = [] 

#Whether or not end devices require an ACK
confirmed_values = [False, True]

#Number of repeats (runs) to do for averaging
runs = 10

# I don't know the exact x-values used, picking my own.
arrival_rates = list(np.logspace(start= -2, stop=1, num=14))

#Fig 2 only plots till 2 pkt/s, but I generated up till 10 pkt/s.
#arrival_rates = arrival_rates[:-3] #they only plot till 2, dropping last 3 generated rates.

num_devices = 1200
for arrival_rate in arrival_rates:

    # 1200 devices will send in 1 appPeriod 1200 packets
    # different appPeriods will generated different arrival rates (lamdba)
    # appPeriod = 1200/(lambda)

    appPeriod_values.append(np.ceil(num_devices/arrival_rate))
    print("For ", np.round(arrival_rate,3), " pkt/s, arrival rate for 1200 devices must be ",  appPeriod_values[-1], "s")





param_combinations = {
    'appPeriod': appPeriod_values,
    'confirmed' : confirmed_values,
}
print(param_combinations)

campaign.run_missing_simulations(param_combinations, runs)

confirmed_flag = False

def get_phy_receivedPackets(result):
    """
    Extract the probability of success from the simulation output.
    From lora-packet-tracker.cc
    Vector packetCounts will contain the following fields: totPacketsSent receivedPackets
    interferedPackets noMoreGwPackets underSensitivityPackets noMoreTx
    """

    outcomes = [float(a) for a in result['output']['stdout'].split()]
    if(confirmed_flag == False):
        return outcomes[1]/outcomes[0] #receivedPackets/totPacketsSent

    if(confirmed_flag == True):
        return outcomes[1]/1200 #outcomes[0] #interferedPackets/totPacketsSent


#Plot unconfirmed 
param_combinations = {
    'appPeriod': appPeriod_values,
    'confirmed' : [False],
}

results_unconfirmed = campaign.get_results_as_xarray(param_combinations,
                                            get_phy_receivedPackets, '', runs)

print("\nResults for get_phy_receivedPackets \n")
results_average = results_unconfirmed.reduce(np.mean, 'runs')
results_std = results_unconfirmed.reduce(np.std, 'runs')

avg = results_average.sel(confirmed = False)
std = results_std.sel(confirmed = False)

print("ULPDR Values for unconfirmed traffic")
ULPDR = np.squeeze(avg).values
for index,item in enumerate(arrival_rates):
    print(np.round(item,3), " pkt/s: ",np.round(ULPDR[index],3))    
      

plt.plot(arrival_rates, np.squeeze(avg),  'ok-' )

#Plot confirmed
confirmed_flag = True

param_combinations = {
    'appPeriod': appPeriod_values,
    'confirmed' : [True],
}

results_confirmed = campaign.get_results_as_xarray(param_combinations,
                                            get_phy_receivedPackets, '', runs)

results_average = results_confirmed.reduce(np.mean, 'runs')
results_std = results_confirmed.reduce(np.std, 'runs')

avg = results_average.sel(confirmed = True)
std = results_std.sel(confirmed = True)
plt.plot(arrival_rates, np.squeeze(avg),'xk-' )

print("ULPDR Values for confirmed traffic")
ULPDR = np.squeeze(avg).values
for index,item in enumerate(arrival_rates):
    print(np.round(item,3), " pkt/s: ",np.round(ULPDR[index],3))    
      
plt.xscale('log')
plt.grid(True, 'both', axis = 'both') #log graph

plt.ylim([0.4,1.01])
plt.xlim([0.01,10.1])
plt.yticks([0.4, 0.45, 0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 1])

plt.legend(['Unconfirmed', 'Confirmed'], title = "Traffic type", loc=(1.01,0.5))
plt.subplots_adjust(right=0.75) #add space to the right so that legend is neatly visible

plt.xlabel("$\lambda$ [pkt/s]")
plt.ylabel("UL-PDR")
plt.title("")

plt.show()
